File: client/parse.go
package client

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"io"
)

// Struct definitions remain the same
type InitData struct {
	Tick    int64  `json:"tick"`
	Text    string `json:"text"`
	Abstick int64  `json:"abstick"`
	Hash    string `json:"hash"`
	Task    string `json:"task"`
}

type NodeData struct {
	Iden     string            `json:"iden"`
	Tags     map[string]string `json:"tags"`
	Props    NodeProps         `json:"props"`
	TagProps map[string]string `json:"tagprops"`
	NodeData map[string]string `json:"nodedata"`
	Path     map[string]string `json:"path"`
}

type NodeProps struct {
	Created int64   `json:".created"`
	Name    string  `json:"name "`
	Seen    []int64 `json:".seen"`
	Type    string  `json:"type"`
	Desc    string  `json:"desc"`
	Leaker  string  `json:"leaker"`
}

type Node struct {
	Key  string     `json:"key"`
	Data [][]string `json:"data"`
	Info NodeData   `json:"info"`
}

type FiniData struct {
	Tock    int64 `json:"tock"`
	Abstock int64 `json:"abstock"`
	Took    int   `json:"took"`
	Count   int   `json:"count"`
}

// ParseJSONStream handles multiple JSON arrays in a stream
func ParseJSONStream(input []byte) ([]InitData, []Node, []FiniData, error) {
	reader := bufio.NewReader(bytes.NewReader(input))

	var initList []InitData
	var nodeList []Node
	var finiList []FiniData

	decoder := json.NewDecoder(reader)

	// Read JSON elements in sequence
	for {
		var data []interface{}
		if err := decoder.Decode(&data); err == io.EOF {
			break
		} else if err != nil {
			return nil, nil, nil, fmt.Errorf("error parsing JSON: %w", err)
		}

		// Process each JSON entry
		switch data[0].(string) {
		case "init":
			var init InitData
			bytes, _ := json.Marshal(data[1])
			json.Unmarshal(bytes, &init)
			initList = append(initList, init)

		case "node":
			node := Node{Key: "node"}
			if nodeArray, ok := data[1].([]interface{}); ok {
				// Extract node pairs
				if nodePairs, ok := nodeArray[0].([]interface{}); ok {
					for _, pair := range nodePairs {
						if pairSlice, ok := pair.([]interface{}); ok {
							var strPair []string
							for _, item := range pairSlice {
								if str, ok := item.(string); ok {
									strPair = append(strPair, str)
								}
							}
							node.Data = append(node.Data, strPair)
						}
					}
				}

				// Extract node properties
				bytes, _ := json.Marshal(nodeArray[1])
				json.Unmarshal(bytes, &node.Info)
			}
			nodeList = append(nodeList, node)

		case "fini":
			var fini FiniData
			bytes, _ := json.Marshal(data[1])
			json.Unmarshal(bytes, &fini)
			finiList = append(finiList, fini)
		}
	}

	return initList, nodeList, finiList, nil
}

File: src/gosynapse/parse.py
from __future__ import annotations

from dataclasses import dataclass
from typing import List, Tuple, Any
import json
import logging

logger = logging.getLogger(__name__)

@dataclass
class InitData:
    tick: int
    text: str
    abstick: int
    hash: str
    task: str

@dataclass
class NodeData:
    iden: str
    tags: dict[str, str]
    props: dict[str, Any]
    tagprops: dict[str, str]
    nodedata: dict[str, str]
    path: dict[str, str]

@dataclass
class Node:
    key: str
    data: List[List[str]]
    info: NodeData

@dataclass
class FiniData:
    tock: int
    abstock: int
    took: int
    count: int


@dataclass
class PrintData:
    mesg: str


def parse_json_stream(raw: bytes) -> Tuple[List[InitData], List[Node], List[FiniData], List[PrintData]]:
    """Parse a stream of JSON messages produced by Synapse.

    Args:
        raw: Raw bytes from the server.

    Returns:
        A tuple of lists: (init messages, nodes, fini messages, print messages).
    """
    text = raw.decode(errors="ignore")
    init_items: List[InitData] = []
    nodes: List[Node] = []
    fini_items: List[FiniData] = []
    print_items: List[PrintData] = []

    for line in text.splitlines():
        line = line.strip()
        if not line:
            continue
        try:
            data = json.loads(line)
        except json.JSONDecodeError:
            logger.debug("Failed to decode JSON line: %s", line)
            continue

        if not isinstance(data, list) or len(data) < 2:
            logger.debug("Unexpected storm message: %s", data)
            continue
        key = data[0]
        payload = data[1]
        if key == "init":
            init_items.append(InitData(**payload))
        elif key == "node":
            node_pairs = []
            if isinstance(payload[0], list):
                for pair in payload[0]:
                    node_pairs.append([str(x) for x in pair])
            info = NodeData(**payload[1])
            nodes.append(Node(key="node", data=node_pairs, info=info))
        elif key == "print":
            print_items.append(PrintData(**payload))
        elif key == "fini":
            fini_items.append(FiniData(**payload))
    return init_items, nodes, fini_items, print_items
